addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,447 @@
+// determined in the Reconcile function
+func (r *PackageInstallStatusReconciler) reconcile(clusterClient client.Client, cluster *clusterapiv1beta1.Cluster, clusterRole ClusterRole) error {
+	log := ctrl.LoggerFrom(r.ctx, constants.ClusterNamespaceLogKey, cluster.Namespace, constants.ClusterNameLogKey, cluster.Name).WithName("PackageInstallStatusController")
+	clusterObjKey := client.ObjectKeyFromObject(cluster)
+
+	clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
+	// ClusterBootstrap resource exists in the management cluster, that's why we need to use local client instead of remote client for fetching it
+	if err := r.Client.Get(r.ctx, clusterObjKey, clusterBootstrap); err != nil {
+		return err
+	}
+	clusterBootstrap = clusterBootstrap.DeepCopy()
+
+	patchHelper, err := clusterapipatchutil.NewHelper(clusterBootstrap, r.Client)
+	if err != nil {
+		return err
+	}
+
+	// this shouldn't include kapp ctrl package as it'll get processed separately
+	packages := append([]*runtanzuv1alpha3.ClusterBootstrapPackage{
+		clusterBootstrap.Spec.CNI,
+		clusterBootstrap.Spec.CPI,
+		clusterBootstrap.Spec.CSI,
+	}, clusterBootstrap.Spec.AdditionalPackages...)
+
+	var cbStatusChanged bool
+
+	for _, pkg := range packages {
+		if pkg == nil {
+			continue
+		}
+		cbConditionChanged, err := r.reconcileClusterBootstrapStatus(clusterClient, &clusterBootstrap.Status, clusterObjKey, clusterRole, pkg.RefName, r.Config.SystemNamespace)
+		cbStatusChanged = cbStatusChanged || cbConditionChanged
+		if err != nil {
+			// in this case, we delete the existing condition corresponding to the package in the ClusterBootstrapStatus (if existing); as the corresponding pkgi or package resources do not exist for the package anymore
+			// then just continue with collecting PackageInstallStatus for other packages
+			cbStatusChanged = r.removeConditionIfExistsForPkgName(&clusterBootstrap.Status, pkg.RefName) || cbStatusChanged
+		}
+	}
+
+	// Note: kapp ctrl pkgi exists only for the workload cluster
+	// it is installed under cluster.Namespace in the management cluster and should be handled separately
+	if clusterRole == clusterRoleWorkload && clusterBootstrap.Spec.Kapp != nil {
+		cbConditionChanged, err := r.reconcileClusterBootstrapStatus(r.Client, &clusterBootstrap.Status, clusterObjKey, clusterRole, clusterBootstrap.Spec.Kapp.RefName, cluster.Namespace)
+		cbStatusChanged = cbStatusChanged || cbConditionChanged
+		if err != nil {
+			// in this case, we delete the existing condition corresponding to the package in the ClusterBootstrapStatus (if existing); as the corresponding pkgi or package resources do not exist for the package anymore
+			// then just proceed with updating the ClusterBootstrapStatus for other packages in one update operation
+			cbStatusChanged = r.removeConditionIfExistsForPkgName(&clusterBootstrap.Status, clusterBootstrap.Spec.Kapp.RefName) || cbStatusChanged
+		}
+	}
+
+	if !cbStatusChanged {
```

Comment:
Absolutely up to you but it seems we only track this for this log line.

The patch helper already internal diffs and decides if there is a need to patch conditions or if there's nothing to do.

Furthermore we're working on improving the logging of patchhelper so patchHelper will internally log in the future if there was a change or not.

So I think you can get rid of the complexity of tracking status changed if you want

Reply:
I got rid of that:)